### PR TITLE
105554146 queeing messages on network unavailable

### PR DESCRIPTION
**Pull request by karthikbr82**

### CODE REVIEW

File: callstats-java-sdk/src/main/java/io/callstats/sdk/BridgeStatusInfoQueue.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		bridgeStatusInfoList.add(bridgeStatusInfo);
+	}
+	
+	public void clear() {
+		bridgeStatusInfoList.clear();
+	}
+	
+	public BridgeStatusInfo pop() {
+		BridgeStatusInfo bridgeStatusInfo = null;
+		if (bridgeStatusInfoList.size() >0) {
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/BridgeStatusInfoQueue.java]: Tab after ">"

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java [hunk 2]
```diff
@@ -148,43 +156,59 @@ public void sendCallStatsBridgeStatusUpdate(BridgeStatusInfo bridgeStatusInfo) {
...
+						} catch (IOException e) {
+							logger.error("IO Execption "+e.getMessage(),e);
+							throw new RuntimeException(e);					
+						} catch (JsonSyntaxException e) {
+							logger.error("Json Syntax Exception "+e.getMessage(),e);
+							e.printStackTrace();
+							throw new RuntimeException(e);
+						}
 						logger.debug("Response status is "+eventResponseMessage.getStatus()+":"+eventResponseMessage.getReason());
 						if (eventResponseMessage.getStatus().equals("Error") && eventResponseMessage.getReason().contains(CallStatsErrors.INVALID_TOKEN_ERROR.getReason())) {
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java]: Replace with enum

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java [hunk 3]
```diff
@@ -148,43 +156,59 @@ public void sendCallStatsBridgeStatusUpdate(BridgeStatusInfo bridgeStatusInfo) {
...
 							//logger.debug("Response status is "+eventResponseMessage.getStatus()+":"+eventResponseMessage.getReason());
 							bridgeKeepAliveManager.stopKeepAliveSender();
 							authenticator.doAuthentication();
-						}						
+						}
+						httpClient.setDisrupted(false);
+						sendCallStatsBridgeStatusUpdateFromQueue();
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java]: Why do we send messages from queue here? Did we get disrupted somewhere?

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java [hunk 4]
```diff
@@ -148,43 +156,59 @@ public void sendCallStatsBridgeStatusUpdate(BridgeStatusInfo bridgeStatusInfo) {
...
+	}
+	
+	
+	public synchronized void sendCallStatsBridgeStatusUpdateFromQueue() {
+		if (bridgeStatusInfoQueue.getLength() < 1)
+			return;
+		
+		while (bridgeStatusInfoQueue.getLength() > 0) {
+			BridgeStatusInfo bridgeStatusInfo = bridgeStatusInfoQueue.pop();
+			sendCallStatsBridgeStatusUpdate(bridgeStatusInfo);
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java]: Do we need any message pacing here? Can we cause message storm in the backend by sending too many events simultaneously?

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsAuthenticator.java [hunk 5]
```diff
@@ -207,24 +192,22 @@ private void handleAuthenticationChallenge(final int appId, final String appSecr
...
+						e.printStackTrace();
+						throw new RuntimeException(e);
+					} catch (IOException e) {
+						e.printStackTrace();
+						throw new RuntimeException(e);					
+					} catch (JsonSyntaxException e) {
+						e.printStackTrace();
+						throw new RuntimeException(e);
+					}
 					if (challengeResponseMessage.getStatus().equals("OK")) {
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsAuthenticator.java]: Enum?

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsAuthenticator.java [hunk 6]
```diff
@@ -239,10 +222,10 @@ public void onResponse(HttpResponse response) {
...
 					}
-				} else if (responseStatus == GATEWAY_ERROR) {
+				} else if (responseStatus == CallStatsResponseStatus.GATEWAY_ERROR ) {
 					scheduleAuthentication(appId, appSecret, bridgeId,httpClient);
 					listener.onError(CallStatsErrors.APP_CONNECTIVITY_ERROR, "SDK Authentication Error");
-				} else if (responseStatus == INVALID_PROTO_FORMAT_ERROR) {
+				} else if (responseStatus == CallStatsResponseStatus.INVALID_PROTO_FORMAT_ERROR) {
 					listener.onError(CallStatsErrors.AUTH_ERROR, "SDK Authentication Error");
 				} else {
 					listener.onError(CallStatsErrors.HTTP_ERROR, "SDK Authentication Error");
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsAuthenticator.java]: Do we give up if HTTP error is met? Code should write e-mail to us and probably keep on trying to reconnect.

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsAuthenticator.java [hunk 7]
```diff
@@ -367,29 +350,26 @@ private void sendAsyncAuthenticationRequest(final int appId, final String appSec
...
+					} catch (IOException e) {
+						isAuthenticationInProgress = false;
+						e.printStackTrace();
+						throw new RuntimeException(e);
+					} catch (JsonSyntaxException e) {
+						isAuthenticationInProgress = false;
+						e.printStackTrace();
+						throw new RuntimeException(e);
+					}
 					if (authorizeResponseMessage.getStatus().equals("OK")) {
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsAuthenticator.java]: Enum

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsBridgeKeepAliveManager.java [hunk 8]
```diff
@@ -129,42 +130,46 @@ public void run() {
 	 * @param token the token
 	 * @param httpClient the http client
 	 */
-	private void sendKeepAliveBridgeMessage(int appId, String bridgeId, String token, CallStatsHttpClient httpClient) {
+	private void sendKeepAliveBridgeMessage(int appId, String bridgeId, String token, final CallStatsHttpClient httpClient) {
 		long epoch = System.currentTimeMillis()/1000;	
 		String apiTS = ""+epoch;
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsBridgeKeepAliveManager.java]: Space around "+"

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsBridgeKeepAliveManager.java [hunk 9]
```diff
@@ -129,42 +130,46 @@ public void run() {
...
+					} catch (IOException e) {
+						e.printStackTrace();
+						throw new RuntimeException(e);					
+					} catch (JsonSyntaxException e) {
+						logger.error("Json Syntax Exception "+e.getMessage(),e);
+						e.printStackTrace();
+						throw new RuntimeException(e);
+					}
 					logger.info("Response status is "+keepAliveResponse.getStatus()+":"+keepAliveResponse.getReason());
 					if (keepAliveResponse.getStatus().equals("Error") && keepAliveResponse.getReason().contains("Invalid client token")) {
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsBridgeKeepAliveManager.java]: Enums

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsHttpClient.java [hunk 10]
```diff
@@ -55,6 +55,16 @@
 	/** The so time out. */
 	private int soTimeOut;
```
Comment (marcinag) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsHttpClient.java]: What is this timeout variable concerning?

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java (hunk 3 above)
Comment (karthikbr82) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java]: I am sending the messages from Queue on every successful transaction. It just checks if any messages are there in queue and sends it.

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java (hunk 4 above)
Comment (karthikbr82) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStats.java]: Message pacing needed but will be done later.

File: callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsHttpClient.java (hunk 10 above)
Comment (karthikbr82) [inline callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsHttpClient.java]: This is auto generated with some plugin, has to be re written once before release.

**Pull request merged**